module/zfs/dsl_dir.c

```diff
@@ -1334,21 +1334,27 @@ dsl_dir_tempreserve_impl(dsl_dir_t *dd, uint64_t asize, boolean_t netfree,
 	/*
 	 * If they are requesting more space, and our current estimate
 	 * is over quota, they get to try again unless the actual
-	 * on-disk is over quota and there are no pending changes (which
-	 * may free up space for us).
+	 * on-disk is over quota and there are no pending changes
+	 * or deferred frees (which may free up space for us).
 	 */
 	if (used_on_disk + est_inflight >= quota) {
-		if (est_inflight > 0 || used_on_disk < quota ||
-		    (retval == ENOSPC && used_on_disk < quota))
-			retval = ERESTART;
+		if (est_inflight > 0 || used_on_disk < quota)
+			retval = SET_ERROR(ERESTART);
+
+		uint64_t deferred = dsl_pool_deferred_space(dd->dd_pool);
+		uint64_t overquota = used_on_disk >= quota ?
+		    used_on_disk - quota : 0;
+		if (retval == ENOSPC && deferred > overquota)
+			retval = SET_ERROR(ERESTART);
```

Comment:
All this section could be inside "else" from above condition. Also it would remove the "used_on_disk >= quota" check.

Reply:
Good point.  I've gone ahead and made the suggested change.  As part of it added an ASSERT to make it clear this code must never underflow and dropped the local variables.  Please take another look.